hub/core/meta/encode/base_encoder.py

```diff
@@ -0,0 +1,177 @@
+from abc import ABC
+from typing import Any, Sequence
+from hub.constants import ENCODING_DTYPE
+import numpy as np
+
+
+class Encoder(ABC):
+    last_index_index: int = -1
+
+    def __init__(self, encoded=None):
+        """Base class for meta encoders. Handles heavy lifting logic for:
+            - Chunk ID encoder
+            - Shape encoder
+            - Byte positions encoder
+
+        Lookup algorithm is essentially the same for all encoders, however the details are different.
+        You can find all of this information in their respective classes.
+
+        Args:
+            encoded (np.ndarray): Encoded state, if None state is empty. Helpful for deserialization. Defaults to None.
+        """
+
+        self._encoded = encoded
+        if self._encoded is None:
+            self._encoded = np.array([], dtype=ENCODING_DTYPE)
+
+    @property
+    def array(self):
+        return self._encoded
+
+    @property
+    def nbytes(self):
+        return self.array.nbytes
+
+    @property
+    def num_samples(self) -> int:
+        if len(self._encoded) == 0:
+            return 0
+        return int(self._encoded[-1, self.last_index_index] + 1)
+
+    def num_samples_at(self, translated_index: int) -> int:
+        """Calculates the number of samples a row in the encoding corresponds to.
+
+        Args:
+            translated_index (int): This index will be used when indexing `self._encoded`.
+
+        Returns:
+            int: Representing the number of samples that a row's derivable value represents.
+        """
+
+        lower_bound = 0
+        if len(self._encoded) > 1 and translated_index > 0:
+            lower_bound = self._encoded[translated_index - 1, self.last_index_index] + 1
+        upper_bound = self._encoded[translated_index, self.last_index_index] + 1
+
+        return int(upper_bound - lower_bound)
+
+    def translate_index(self, local_sample_index: int) -> int:
+        """Translates `local_sample_index` into which row it corresponds to inside the encoded state.
```

Comment:
These docstrings could use another pass. "Translated" doesn't really tell me anything about the value.